feat(event-handler): add support for Pydantic Field discriminator in validation

dap0am opened the pull request:
## Summary

This PR adds support for Pydantic's `Field(discriminator='...')` syntax in event handler validation, enabling tagged unions to work correctly with the validation middleware.

Fixes #5953

## Changes

- Enhanced `get_field_info_annotated_type()` in `params.py` to handle `Field(discriminator) + Body()` combinations
- Modified `ModelField.__post_init__()` in `compat.py` to preserve discriminator metadata when creating TypeAdapters
- Added comprehensive tests for discriminator validation and other Field features

## Testing

- Added `test_field_discriminator_validation()` to verify discriminator functionality with tagged unions
- Added `test_field_other_features_still_work()` to ensure other Field features remain functional
- All existing tests pass without modification

## Example Usage

```python
from typing import Literal
from typing_extensions import Annotated
from pydantic import BaseModel, Field
from aws_lambda_powertools.event_handler import APIGatewayRestResolver
from aws_lambda_powertools.event_handler.openapi.params import Body

app = APIGatewayRestResolver(enable_validation=True)

class FooAction(BaseModel):
    action: Literal["foo"]
    foo_data: str

class BarAction(BaseModel):
    action: Literal["bar"]
    bar_data: int

# Now works with Field discriminator
Action = Annotated[FooAction | BarAction, Field(discriminator="action")]

@app.post("/actions")
def create_action(action: Annotated[Action, Body()]):
    return {"received": action.model_dump()}
```

## Checklist

- [x] Tests are passing
- [x] Linting and type checking pass
- [x] Follows project contribution guidelines
- [x] Maintains backward compatibility

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

dap0am committed:
feat(event-handler): add support for Pydantic Field discriminator in validation (#5953)

Enable use of Field(discriminator='...') with tagged unions in event handler validation.
This allows developers to use Pydantic's native discriminator syntax instead of
requiring Powertools-specific Param annotations.

- Handle Field(discriminator) + Body() combination in get_field_info_annotated_type
- Preserve discriminator metadata when creating TypeAdapter in ModelField
- Add comprehensive tests for discriminator validation and Field features

dap0am committed:
style(tests): remove inline comments to match project test style

leandrodamascena committed:
Merge branch 'develop' into develop

leandrodamascena commented:
Hey @dap0am please run `make format` to fix the CI.

dap0am committed:
style: run make format to fix CI formatting issues

dap0am commented:
> Hey @dap0am please run `make format` to fix the CI.

Done, thanks.

leandrodamascena commented:
> > Hey @dap0am please run `make format` to fix the CI.
> 
> Done, thanks.

Thanks for fixing! You can run `make pr` before commit the code. So, you can see if there are errors in the code you refactored.

leandrodamascena reviewed (changes requested):
Hey @dap0am! Please let me know if you need any help here.

leandrodamascena committed:
Merge branch 'develop' into develop

dap0am commented:
> Hey @dap0am! Please let me know if you need any help here.

Hi @leandrodamascena nope, I am just waiting for your review.

leandrodamascena commented:
> > > Hey @dap0am please run `make format` to fix the CI.
> > 
> > 
> > Done, thanks.
> 
> Thanks for fixing! You can run `make pr` before commit the code. So, you can see if there are errors in the code you refactored.

Hi @dap0am, can you fix the CI and make the appropriate changes so as not to break the current tests? The CI is failing because there are some refactorings that can introduce bug regressions into the existing code and break all customers using it.

Use `make pr` locally to see which tests are failing and what you need to change in your code to avoid breaking the current tests.

leandrodamascena reviewed (changes requested):
@dap0am please fix the CI before we start the review.

dap0am commented:
> > > > Hey @dap0am please run `make format` to fix the CI.
> > > 
> > > 
> > > Done, thanks.
> > 
> > 
> > Thanks for fixing! You can run `make pr` before commit the code. So, you can see if there are errors in the code you refactored.
> 
> Hi @dap0am, can you fix the CI and make the appropriate changes so as not to break the current tests? The CI is failing because there are some refactorings that can introduce bug regressions into the existing code and break all customers using it.
> 
> Use `make pr` locally to see which tests are failing and what you need to change in your code to avoid breaking the current tests.

Will do that, thanks.

dap0am committed:
fix(event-handler): preserve FieldInfo subclass types in copy_field_info

Fix regression where copy_field_info was losing custom FieldInfo subclass
types (Body, Query, etc.) by using shallow copy instead of from_annotation.
This resolves the failing test_validate_embed_body_param while maintaining
the discriminator functionality.

dreamorosi committed:
Merge branch 'develop' into develop

dreamorosi commented:
Hi @dap0am, some of the unit tests are failing - please take a look if you can and also resolve the SonarCloud issues listed in the comment above.

Thanks!

dap0am committed:
refactor(event-handler): reduce cognitive complexity and address SonarCloud issues

- Refactor get_field_info_annotated_type function by extracting helper functions to reduce cognitive complexity from 29 to below 15
- Fix copy_field_info to preserve FieldInfo subclass types using shallow copy instead of from_annotation
- Rename variable Action to action_type to follow Python naming conventions
- Resolve failing test_validate_embed_body_param by maintaining Body parameter type recognition
- Add helper functions: _has_discriminator, _handle_discriminator_with_body, _create_field_info, _set_field_default
- Maintain full backward compatibility and discriminator functionality

leandrodamascena committed:
Merge branch 'develop' into develop

dap0am committed:
style: fix formatting to pass CI format check

Apply ruff formatting to params.py to resolve failing format check in CI

dap0am commented:
@leandrodamascena @dreamorosi I have fixed the last issue causing the tests to fail.

leandrodamascena committed:
Merge branch 'develop' into develop

dap0am committed:
fix: resolve mypy type error in _create_field_info function

Add explicit type annotation for field_info variable to fix mypy error
about incompatible types between FieldInfo and Body. This ensures type
checking passes across all Python versions (3.9-3.13).

dreamorosi committed:
Merge branch 'develop' into develop

dreamorosi commented:
Hi @dap0am, the tests are failing, so we're unable to merge this.

Please make sure to run the tests as discussed above before pushing code. 

We'll try a couple more iterations, but if we're unable to get to a green CI and no SonarCloud issues we might close the PR.

dap0am commented:
> Hi @dap0am, the tests are failing, so we're unable to merge this.
> 
> Please make sure to run the tests as discussed above before pushing code.
> 
> We'll try a couple more iterations, but if we're unable to get to a green CI and no SonarCloud issues we might close the PR.

Thanks for your comment, I was having issue running the full test suite locally, but I think I have fixed it now, as my test are now passing.

dap0am committed:
fix: use Union syntax for Python 3.9 compatibility

dreamorosi commented:
> I was having issue running the full test suite locally, but I think I have fixed it now, as my test are now passing.

Thanks! If there's anything we can improve to make it easier for contributors to run it, please let us know!

leandrodamascena committed:
Merge branch 'develop' into develop

leandrodamascena commented on aws_lambda_powertools/event_handler/openapi/compat.py:
```diff
@@ -176,7 +186,13 @@ def model_rebuild(model: type[BaseModel]) -> None:
 def copy_field_info(*, field_info: FieldInfo, annotation: Any) -> FieldInfo:
-    return type(field_info).from_annotation(annotation)
+    # Create a shallow copy of the field_info to preserve its type and all attributes
+    from copy import copy
```
Please move this import to top level.

leandrodamascena reviewed (changes requested):
Hey @dap0am thanks for working on this! Overall it looks good to me and the code is working as expected. I see Codecov is complaining about missing coverage and I'll take a look. Also, we need to update the documentation. Do you want to update or do you want me to update the documentation to include we support this new field?

Thanls

dap0am commented:
> Hey @dap0am thanks for working on this! Overall it looks good to me and the code is working as expected. I see Codecov is complaining about missing coverage and I'll take a look. Also, we need to update the documentation. Do you want to update or do you want me to update the documentation to include we support this new field?
> 
> Thanls

I can put something in the documentation.

leandrodamascena commented:
> > Hey @dap0am thanks for working on this! Overall it looks good to me and the code is working as expected. I see Codecov is complaining about missing coverage and I'll take a look. Also, we need to update the documentation. Do you want to update or do you want me to update the documentation to include we support this new field?
> > Thanls
> 
> I can put something in the documentation.

Thanks! Please let me know if need help.

leandrodamascena committed:
Merge branch 'develop' into develop

dap0am committed:
feat(event-handler): add documentation and example for Field discriminator support

leandrodamascena committed:
Merge branch 'develop' into develop

leandrodamascena reviewed (changes requested):
CI is failing @dap0am!

dap0am committed:
style: run make format to fix CI formatting issues

dap0am commented:
> CI is failing @dap0am!

Format issue in the example file, I have fixed it now and passed the test locally.

dap0am commented:
@leandrodamascena CI passed now, aside the codecov issue you mentioned to look into, let me know when you review my changes.

leandrodamascena committed:
Merge branch 'develop' into develop

leandrodamascena committed:
small changes

leandrodamascena committed:
Merge branch 'develop' into develop

leandrodamascena reviewed (dismissed):
Hey @dap0am thanks for working on this! I just made some small changes in methods names + documentation and we are good to merge!.

APPROVED! 🚀

leandrodamascena committed:
small changes

dap0am commented:
> Hey @dap0am thanks for working on this! I just made some small changes in methods names + documentation and we are good to merge!.
> 
> APPROVED! 🚀

Thanks @leandrodamascena and @dreamorosi

hjgraca reviewed (approved):
GTM
